### PR TITLE
Updated input op tracking software used for

### PR DESCRIPTION
**Pull request by alexmsettle**
building the network graph.
nvmarker.py and dlprof.py use callid as a unique id for
each operator.  input_callid is the unique id list of the
input callids to each operator.  These fields are included in the
NVTX_EVENTS table in the sqlite file.  The callid can be used
in conjunction with the funcStack to determine a unique descriptive
name for each operator.

Added new functions that are exported in __init__.py so users
can call them from their instrumentation code.

debug() enables/disables printf debug messages
start_graph() - starts network graph collection
stop_graph() - stops network graph collection

New config option --delay_graph_capture - defaults to False.
   User sets this to True when they want to use start/stop graph
   to control when to collect the network graph information.

Signed-off-by: Alex Settle <alexmsettle@gmail.com>

### CODE REVIEW

File: pyprof/nvtx/nvmarker.py
Comment (debermudez) [inline pyprof/nvtx/nvmarker.py:65]: fstring

File: pyprof/nvtx/nvmarker.py
Comment (debermudez) [inline pyprof/nvtx/nvmarker.py:77]: fstring

File: pyprof/nvtx/nvmarker.py
Comment (debermudez) [inline pyprof/nvtx/nvmarker.py:54]: Is there a way to avoid using a global here?

File: pyprof/nvtx/nvmarker.py
Comment (debermudez) [inline pyprof/nvtx/nvmarker.py:207]: fstring

File: pyprof/nvtx/nvmarker.py
Comment (debermudez) [inline pyprof/nvtx/nvmarker.py:267]: fstring

**Review by debermudez — CHANGES_REQUESTED** [review]

**Comment by adityaiitb** [comment]
Can we move all the graph related functions and utilities in a separate class / file and make it opt in. We should refactor the nvtx directory so that its easier to integrate stubs (e.g. the current contribution by Alex).

**Comment by debermudez** [comment]
Is that a general request or something we should do as part of this PR?

**Comment by adityaiitb** [comment]
The first comment was for this PR. The functions `start_graph`, `stop_graph`, `debug`, `dprint` are specific to this contribution and so it would be nice if they are in a separate class. The `debug` function is actually not necessary. It can be an argument to `start_graph`, for example

```
start_graph(debug=False)
stop_graph()
```

as opposed to
```
debug()
start_graph()
stop_graph()
```

The second comment was more generic.

**Comment by debermudez** [comment]
okay thanks for the clarification. 
do you have an idea for how to do that refactor?

**Review by debermudez — APPROVED** [review]

**Comment by debermudez** [comment]
@adityaiitb this looks good to me. Good with you?

File: pyprof/nvtx/nvmarker.py
Comment (adityaiitb) [inline pyprof/nvtx/nvmarker.py:203]: 1. The `wrapper_func` is in the critical path. The way the function `dprint` is defined and called here will cause a very expensive string formatting operation irrespective of whether the string is printed or not. Given that in most cases debug is False, the formatted string will end up not being printed. This will slow down profiling significantly.
2. A lot of activity happens inside the function `wrapper_func` and so it would be nice to move such code to a helper function.

File: pyprof/nvtx/nvmarker.py
Comment (adityaiitb) [inline pyprof/nvtx/nvmarker.py:260]: Same as above. Move to a helper function and optimize code so that when debug is False, this function is almost a noop.

File: pyprof/nvtx/nvmarker.py
Comment (adityaiitb) [inline pyprof/nvtx/nvmarker.py:265]: Same as above. Move to a helper function and optimize code so that when debug is False, this function is almost a noop.

File: pyprof/nvtx/nvmarker.py
Comment (adityaiitb) [inline pyprof/nvtx/nvmarker.py:75]: I am not sure, but these functions appear to be dlprof specific (?) and so they should be part of that class.

**Comment by adityaiitb** [comment]
@alexmsettle: If the functionality of the current PR can be made independent of pyprof / dlprof, please consider making a separate class altogether. I imagine you have separate scripts to process the resulting sqlite database which is independent of pyprof / dlprof?

The scope of `nvmarker.py` (monkey patching) has grown. We are not there yet but we are trying to simplify and make it more modular. I believe, there is some code reuse between the current PR and dlprof code. In that case, we should pull out the common code to some utility functions as opposed to munging your contribution with dlprof code, unless that was the intent.

File: pyprof/nvtx/config.py
Comment (adityaiitb) [inline pyprof/nvtx/config.py:30]: Can we rename this flag to something like `debug_graph`? We don't yet have it but its a good idea to have separate debug flags for separate features.

File: pyprof/nvtx/dlprof.py
Comment (alexmsettle) [reply pyprof/nvtx/dlprof.py]: Ok I fixed this in last update

File: pyprof/nvtx/dlprof.py
Comment (alexmsettle) [reply pyprof/nvtx/dlprof.py]: I changed all the .formats to f string

File: pyprof/nvtx/nvmarker.py
Comment (alexmsettle) [reply pyprof/nvtx/nvmarker.py]: fixed this too

File: pyprof/nvtx/nvmarker.py
Comment (alexmsettle) [reply pyprof/nvtx/nvmarker.py]: done

File: pyprof/nvtx/nvmarker.py
Comment (alexmsettle) [reply pyprof/nvtx/nvmarker.py]: moved this to the dlprof class

File: pyprof/nvtx/nvmarker.py
Comment (alexmsettle) [reply pyprof/nvtx/nvmarker.py]: done

File: pyprof/nvtx/nvmarker.py
Comment (alexmsettle) [reply pyprof/nvtx/nvmarker.py]: fixed

File: pyprof/nvtx/config.py
Comment (alexmsettle) [reply pyprof/nvtx/config.py:30]: done debug_graph

File: pyprof/nvtx/nvmarker.py
Comment (alexmsettle) [reply pyprof/nvtx/nvmarker.py]: I just removed the dprints.

Are you saying the *arg parameter that is passed to dprint is evaluated every time regardless of whether debug is set?   I have not noticed a performance issue with this.

**Review by alexmsettle — COMMENTED** [review]
I think I covered all the items called out in the code review.

I did not create a new class for start_graph() / stop_graph().
I suppose they could be static member functions in the dlprof class, I don't think it would make much of a difference either way

**Comment by debermudez** [comment]
@adityaiitb ping. Waiting on your approval before i merge.

**Review by debermudez — APPROVED** [review]

**Pull request merged**
